src/pce/pce.jl

```diff
@@ -0,0 +1,497 @@
+    univariate orthogonal polynomial basis ``\\{ψ_{α_i}^{(i)}(x_i)\\}_{α_i=0}^{r_i}`` for
+    each ``X_i`` and multi-indices ``α⃗``.
+    """
+    tensor_basis::TensorProductOrthoPoly
+    "The coefficients ``C_{(α_1, α_2, …, α_n)}`` for each ``Y_i`` in the columns."
+    moments::Matrix
+    "The mapping ``Y => ∑_α C_α Ψ_α`` for each ``Y``."
+    ansatz::Dict{Num, Num}
+    "Results of tensor inner products ``⟨Ψ_{i_1}Ψ_{i_2}⋯Ψ_{i_{m-1}},Ψ_{i_m}⟩``."
+    tensors::Dict{Int, Tensor}
+end
+"""
+$(TYPEDSIGNATURES)
+Construct a `$(FUNCTIONNAME)` object.
+
+In practice, the infinite series of multi-indices must be truncated. By default, besides
+the upper bound for the degree of each univariate orthogonal polynomial, the total degree
+is restricted to the maximum degree among the univariate bases. That is
+```math
+\\begin{gather*}
+α_1 + α_2 + ⋯ α_n ≤ \\max_i r_i \\\\
+0 ≤ α_i ≤ r_i \\quad ∀ i
+\\end{gather*}
+```
+where ``r_i`` is the degree upper bound for the univariate basis corresponding to ``X_i``.
+
+# Arguments
+- `states`: Random vairables ``\\mathbf Y`` that are represented by other random variables.
+- `ivs`: Independent vairables of ``\\mathbf Y``. Enter an empty vector if there is none.
+- `parameters`: Independent random variables ``\\mathbf X``.
+- `uni_basis`: Univariate orthogonal polynomial basis ``\\{ψ_{α_i}^{(i)}(x_i)\\}_{α_i=0}^{r_i}`` for each ``X_i``.
+"""
+function PCE(states::AbstractVector{Num}, ivs::AbstractVector{Num},
+             parameters::AbstractVector{Num},
+             uni_basis::AbstractVector{
+                                       <:Union{AbstractOrthoPoly, AbstractCanonicalOrthoPoly
+                                               }})
+    states = Symbolics.scalarize(states)
+    parameters = Symbolics.scalarize(parameters)
+    tensor_basis = TensorProductOrthoPoly(uni_basis)
+    C = Matrix{Num}(undef, dim(tensor_basis), length(states))
+    Ψ = Vector{Num}(undef, dim(tensor_basis))
+    snames = Symbolics.tosymbol.(states)
+    for i in axes(tensor_basis.ind, 2)
+        name = join(Symbolics.map_subscripts.(view(tensor_basis.ind, :, i)), "ˏ")
+        Ψname = Symbol(:Ψ, name)
+        # record index of multi-index α⃗ in symbolic metadata
+        Ψ[i] = first(@variables $Ψname [description = i])
+        for (j, sname) in enumerate(snames)
+            Cname = Symbol(:C, sname, name)
+            C[i, j] = if isempty(ivs)
+                first(@variables $Cname [description = i])
+            else
+                first(@variables $Cname(ivs...) [description = i])
+            end
+        end
+    end
+    # X => x₀ + x₁ψ₁(X)
+    # where x₀ and x₁ are affince PCE coefficients of X
+    # ψ₁(x) = x - α₀ is the first-order monic basis polynomial of X
+    # TODO: Use mean and std of the distribution of X to compute x₀ and x₁
+    x_dict = Dict(x => op.α[1] + Ψ[i + 1]
+                  for (i, (x, op)) in enumerate(zip(parameters, uni_basis)))
```

Comment:
Create a struct for otho poly basis $Ψ_α$ (something similar to `SemiMonomial` in `Symbolics.semipolynomial_form`)
Use term rewriter to deal with differential operator `@rule D(CΨ) => Ψ*D(C)` and evaluate tensor scalar products in the tree `@rule *(~~Ψ) => something`.